src/libraries/System.Security.Cryptography.X509Certificates/tests/ExtensionsTests/SubjectAlternativeNameTests.cs

```diff
@@ -0,0 +1,149 @@
+            X509Extension built = builder.Build(true);
+
+            X509SubjectAlternativeNameExtension ext;
+
+            switch (loadMode)
+            {
+                case LoadMode.CopyFrom:
+                    ext = new X509SubjectAlternativeNameExtension();
+                    ext.CopyFrom(built);
+                    break;
+                case LoadMode.Array:
+                    ext = new X509SubjectAlternativeNameExtension(built.RawData);
+                    break;
+                case LoadMode.Span:
+                    byte[] tmp = new byte[built.RawData.Length + 2];
+                    built.RawData.AsSpan().CopyTo(tmp.AsSpan(1));
+                    ext = new X509SubjectAlternativeNameExtension(tmp.AsSpan()[1..^1]);
+                    tmp.AsSpan().Clear();
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(loadMode), loadMode, "Unexpected mode");
+            }
+
+            Assert.Equal(new[] { IPAddress.Loopback, IPAddress.IPv6Loopback }, ext.EnumerateIPAddresses());
+        }
+
+        [Theory]
+        [InlineData(LoadMode.CopyFrom)]
+        [InlineData(LoadMode.Array)]
+        [InlineData(LoadMode.Span)]
+        public static void VerifyInvalidDataBehavior(LoadMode loadMode)
+        {
+            byte[] invalidEncoding = { 0x05, 0x00 };
+
+            switch (loadMode)
+            {
+                case LoadMode.CopyFrom:
+                    X509Extension untyped = new X509Extension("0.0", invalidEncoding, true);
+                    X509SubjectAlternativeNameExtension ext = new X509SubjectAlternativeNameExtension();
+
+                    // The pattern for X509Extension is that CopyFrom doesn't validate data,
+                    // and it blindly accepts the incoming OID.  The semantic properties then throw late.
+                    ext.CopyFrom(untyped);
+                    Assert.True(ext.Critical);
+                    Assert.Equal("0.0", ext.Oid.Value);
+                    AssertExtensions.SequenceEqual(invalidEncoding, ext.RawData);
+                    Assert.Throws<CryptographicException>(ext.EnumerateDnsNames);
+                    Assert.Throws<CryptographicException>(ext.EnumerateIPAddresses);
+                    break;
+                case LoadMode.Array:
+                    // The ctors don't need to be so forgiving, through.
+                    Assert.Throws<CryptographicException>(
+                        () => new X509SubjectAlternativeNameExtension(invalidEncoding));
+                    break;
+                case LoadMode.Span:
+                    Assert.Throws<CryptographicException>(
+                        () => new X509SubjectAlternativeNameExtension(new ReadOnlySpan<byte>(invalidEncoding)));
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(loadMode), loadMode, "Unexpected mode");
+            }
+        }
```

Comment:
What do you think about tests for invalid data for a DNSName or IPAddress?

1. IPAddress that is an OCTET STRING that is not 4 or 6 bytes in length. (Looking for a test for the `Cryptography_X509_SAN_UnknownIPAddressSize` error)
2. A DNSName that is not an IA5String.

Would that lazily throw during enumeration, or is it validated up-front? I don't have strong opinions one way or the other what it does, but I'd be interested to know the behavior.

Reply:
Apparently it was the case an invalid IA5 was a Decode failure and bad ipAddress length was a late failure.  Made them both be failures in Decode